src/lib/SX1280Driver/SX1280.cpp

```diff
@@ -579,21 +583,102 @@ int8_t ICACHE_RAM_ATTR SX1280Driver::GetRssiInst(SX12XX_Radio_Number_t radioNumb
+                WORD_ALIGNED_ATTR uint8_t RXdataBuffer_second[RXBuffSize];
+                hal.ReadBuffer (FIFOaddr, RXdataBuffer_second, PayloadLength, radio[secondRadioIdx]);
+
+                // leaving only the type in the first byte (crcHigh was cleared)
+                RXdataBuffer[0] &= 0b11;
+                RXdataBuffer_second[0] &= 0b11;
+                // if the second packet is same to the first, it's valid
+                if(memcmp(RXdataBuffer, RXdataBuffer_second, PayloadLength) == 0)
+                    isSecondRadioGotData = true;
+            }
+        }
+
+        // second radio received the same packet to the processing radio
+        gotRadio[secondRadioIdx] = isSecondRadioGotData;
+    }
+
     uint8_t status[2];
-    hal.ReadCommand(SX1280_RADIO_GET_PACKETSTATUS, status, 2, processingPacketRadio);
-    if (packet_mode == SX1280_PACKET_TYPE_FLRC) {
-        // No SNR in FLRC mode
-        LastPacketRSSI = -(int8_t)(status[1] / 2);
-        LastPacketSNRRaw = 0;
-        return;
+    uint8_t rssi[2];
+    uint8_t snr[2];
+
+    for(uint8_t i=0;i<2;i++)
+    {
+        if (gotRadio[i])
+        {
+            hal.ReadCommand(SX1280_RADIO_GET_PACKETSTATUS, status, 2, radio[i]);
+
+            if (packet_mode == SX1280_PACKET_TYPE_FLRC) {
+                // No SNR in FLRC mode
+                rssi[i] = -(int8_t)(status[1] / 2);
+                snr[i] = 0;
+            }
+            else
+            {
+                // LoRa mode has both RSSI and SNR
+                rssi[i] = -(int8_t)(status[0] / 2);
+                snr[i] = (int8_t)status[1];
+
+                // https://www.mouser.com/datasheet/2/761/DS_SX1280-1_V2.2-1511144.pdf p84
+                // need to subtract SNR from RSSI when SNR <= 0;
+                int8_t negOffset = (snr[i] < 0) ? (snr[i] / RADIO_SNR_SCALE) : 0;
+                rssi[i] += negOffset;
+            }
+
+            // If radio # is 0, update LastPacketRSSI, otherwise LastPacketRSSI2
+            (i == 0) ? LastPacketRSSI = rssi[i] : LastPacketRSSI2 = rssi[i];
+            // Update whatever SNRs we have
+            LastPacketSNRRaw = snr[i];
+        }
     }
-    // LoRa mode has both RSSI and SNR
-    LastPacketRSSI = -(int8_t)(status[0] / 2);
-    LastPacketSNRRaw = (int8_t)status[1];
-    // https://www.mouser.com/datasheet/2/761/DS_SX1280-1_V2.2-1511144.pdf p84
-    // need to subtract SNR from RSSI when SNR <= 0;
-    int8_t negOffset = (LastPacketSNRRaw < 0) ? (LastPacketSNRRaw / RADIO_SNR_SCALE) : 0;
-    LastPacketRSSI += negOffset;
+
+    // by default, set the last successful packet radio to be the current processing radio (which got a successful packet)
+    instance->lastSuccessfulPacketRadio = instance->processingPacketRadio;
+
+    // when both radio got the packet, use the better RSSI one
+    if(gotRadio[0] && gotRadio[1])  {
+        // LastPacketSNRRaw = (snr[0]>snr[1])? snr[0]: snr[1]; // design choice
+        LastPacketSNRRaw = (snr[0]+snr[1])/2;
```

Comment:
This was discussed this a while back on the discord here : https://discord.com/channels/596350022191415318/733424557636976692/1084858664088784907

A nice option would be to be able to choose between weakest/strongest (or averaged) link for the dynamic radio.
Those could be named Diversity-Reliable, Diversity-PowerSave as Sunjun proposed.

It may not be required to climb the power each time because of the weakest link.

This is only the RSSi graph, but the difference between the 2 radios can be quite important:
I'll try to do a test with both SNR instead of both RSSI (by changing the values transmitted).
![image](https://user-images.githubusercontent.com/182152/228387895-1123187e-16fb-43f0-a5f3-fccb3bfc63d9.png)

Reply:
_Diversity-Reliable, Diversity-PowerSave_ 

If we give the user this option I know I would select the Reliable option.  If you make PowerSave the default you are now making a diversity rx appear to under perform to the user.

Im also conscious of adding another option when ELRS appears to be complicated.

Reply:
Personally I think the idea of dual radio is to improve the performance and reliability, so why would be ever choose the best SNR for driving DP! We should always choose the worst so that we bump the power to maintain the link with the best possible reliability level. Of course, this has it's own set of problems i.e. if one of the antennas gets eaten by the props then we are going to cause the TX to beef up the power to compensate for that, but maybe that's ok.

Reply:
This already happens if a user is mountain surfing with a mono antenna rx and turns in such a way that the antenna is cross polarized.

Diversity receivers should have a clear benefit over mono radio receivers.  Not be restricted to only marginally better.

Reply:
Again this is a design choice of defining how our diversity and DP works.

This is my opinion: the advantage of a true diversity receiver is that it allows for at least one packet to be received by one of the two antennas, which prevents a failsafe situation. As long as one antenna reliably receives packets, the other antenna doesn't necessarily need to receive packets all the time.

However, Jye and PK have a different approach. According to them, both antennas should always have the maximum chance of receiving a packet. This approach is quite different, as it aims for a constant LQ of 100 rather than simply preventing failsafe situations.

Reply:
Also.. just for your information.

RSSI dynamic power will inc/dec power on the signal level of the stronger radio with this PR as well.

`dynpower.cpp L64`
`int8_t rssi = (CRSF::LinkStatistics.active_antenna == 0) ? CRSF::LinkStatistics.uplink_RSSI_1 : CRSF::LinkStatistics.uplink_RSSI_2;`

In this PR `active_antenna` is always set to the stronger RSSI one.

Reply:
> This already happens if a user is mountain surfing with a mono antenna rx and turns in such a way that the antenna is cross polarized.

In this case, diversity receivers will get the packet from the other antenna? because one antenna gets out of polarization, the other will be aligned naturally? Why do you think it isn't the benefit over a monopole?

Reply:
I added a discussion flag to this PR, so everyone please join and leave your opinion how it should work

Reply:
The implementation was updated to use fuzzy logic, working as follows:
When the difference between those two SNRs is larger than a threshold, it will return the averaged value between the two. When the difference between the two values is small, return the lower SNR value. On the borderline differences, the transition between the two logics is smoothened by fuzzy logic.

The threshold of the fuzzy logic is determined by dynamic power up/down SNR thresholds.

The following image presents how the logic works:
X: SNR difference
Y: high/low/averaged/fuzzy SNR values

![image](https://user-images.githubusercontent.com/12195507/230089634-08216604-adaa-4775-85fe-10184caf4389.png)